### PR TITLE
[I-59] deployment state - feature

### PR DESCRIPTION
**Pull request by skcc321**
# What does it do?
add the feature https://github.com/skcc321/deploy_pin/issues/59
- support for `negative` identifiers (those tasks are going to be executed in the reverse order from the)
- deployment_state tracking: `DeployPin.deployment_ongoing?`
- recurring tasks (id:group:recurring)
- bump gh actions

a new tasks sequence looks like this:
image

image

image

image

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -187,6 +187,29 @@ DeployPin::Database::execute_with_timeout do
...
 ```
+## DeploymentStateTrack
+in the initializer
+```ruby
+    DeployPin.setup do
+      groups %w[I II III post rollback]
+      ...
+      deployment_state_transition({
+        ongoing: %w[I III],
+        pending: "rollback", # enters to pending stet before "rollback"
```
Comment (dovganych) [inline README.md:198]: type in comment
```suggestion
        pending: "rollback", # enters to pending step before "rollback"
```

File: lib/deploy_pin/deployment_state.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    DEPLOYMENT_PENDING_TASK = %(
+      # no_file_task
+      # 0:%<group>s:recurring
+      # task_title: Set DeployPin.state to 'pending'
+
+      DeployPin.send(:deployment_pending!)
+      ).strip
+
+    def ongoing_deployment?
```
Comment (dovganych) [inline lib/deploy_pin/deployment_state.rb:35]: is it possible to make deployment_state an enumerator?

File: lib/deploy_pin/deployment_state.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        in_memory_store.delete(:deployment)
+        Rails.cache.delete(CACHE_KEY)
+      end
+
+      def deployment_ongoing!
+        Rails.cache.write(CACHE_KEY, ONGOING)
+      end
+
+      def deployment_pending!
+        Rails.cache.write(CACHE_KEY, PENDING)
```
Comment (ybregey) [inline lib/deploy_pin/deployment_state.rb:68]: probably OK for now, though it would be nice to pass cache storage via config.

Comment (skcc321) [reply lib/deploy_pin/deployment_state.rb:68]: good point! Thank you @ybregey, will do that.

File: lib/deploy_pin/deployment_state.rb (hunk 2 above)
Comment (skcc321) [reply lib/deploy_pin/deployment_state.rb:65]: What is the purpose for that?

**Review by dovganych — APPROVED** [review]

**Review by okosyk — APPROVED** [review]

**Review by serhii-bodnaruk — APPROVED** [review]

**Review by dkh — APPROVED** [review]

**Review by SviridovSV — APPROVED** [review]

**Review by ochupa — APPROVED** [review]

**Review by ybregey — APPROVED** [review]

**Pull request merged**